### PR TITLE
Add tsNodeOpts to enable configuring auto loaded ts-node

### PR DESCRIPTION
**Pull request by r4j4h**
## Proposed changes

https://github.com/webdriverio/webdriverio/pull/6275 introduced automatic loading of ts-node. I believe that this was accidentally a breaking change for users who need to configure ts-node differently than the automatic loader. The proposed workaround is to remove custom loading code and rely on the automatic loader, which may not work for all users. Since webdriverio worked for them before the change, it should work them after.

I think we need the loadTypeScriptCompiler mechanism to look through the config file for something like a tsNodeOpts block and spread that with its defaults. The defaults can be the current value so that users who do not define it get the current behavior. Users who need to modify it could now modify their wdio.conf.js file as neede, remove their custom loading code, have everything working well.

There are more details described in my response to pull 6275: https://github.com/webdriverio/webdriverio/pull/6275#issuecomment-770051932

My change ended up being a little more expansive, in order to provide the config I had to reorder where loadTypeScriptCompiler is called in ConfigParser. I also updated the types.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

I am a little hesitant about the documentation part, otherwise I would love to add some. I tried to add it to the examples.

Please let me know if I missed any areas.

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [X] I have added proper type definitions for new commands (if appropriate)

## Further comments

- I refactored the non-loading logic from  ConfigParser's addConfigFile into addConfigEntry so that I could spread a modified config with tsConfigOpts in the test. This was not necessary, but would have required me adding another config file to the fixtures solely for this purpose and I did not want to muddy the waters especially in my first commit. Besides I do think this separates the logic a little more cleanly.

- I used ts-node's RegisterOptions type in packages/wdio-config/src/utils.ts but this could just as well be any. I call this out incase it might unknowingly make ts-node's types a dependency. If so let me know and I will set it to any.

### CODE REVIEW

File: packages/wdio-config/src/lib/ConfigParser.ts
Comment (christian-bromann) [inline packages/wdio-config/src/lib/ConfigParser.ts:93]: This should happen before we call `addConfigEntry` otherwise any imports or TS code in the config can't be compiled, no?

File: packages/wdio-types/src/index.ts
Comment (christian-bromann) [inline packages/wdio-types/src/index.ts:24]: We can move this into the options.ts file and import it in the files where we need it directly. These options are only defined here so we can overwrite them in case someone has `@wdio/cucumber-framework` etc installed where we have access to these types.
```suggestion
```

File: packages/wdio-types/src/Options.ts
Comment (christian-bromann) [inline packages/wdio-types/src/Options.ts:406]:
```suggestion
    tsNodeOpts?: { [key: string]: any }
```

File: packages/wdio-config/src/utils.ts
Comment (christian-bromann) [inline packages/wdio-config/src/utils.ts:224]:
```suggestion
export function loadTypeScriptCompiler (tsNodeOpts?: RegisterOptions) {
```

**Review by christian-bromann — COMMENTED** [review]
Great job! I was already thinking about this the other day. This is very needed.

File: packages/wdio-config/src/lib/ConfigParser.ts
Comment (r4j4h) [reply packages/wdio-config/src/lib/ConfigParser.ts:93]: Great question! I moved it here so that we could have the options parsed to hand to ts-node. Is there a better spot off-hand we could put it, or do we need to minimally open the config to get the options and then re-parse after?

Comment (christian-bromann) [reply packages/wdio-config/src/lib/ConfigParser.ts:93]: Ah yes, we can't read the `tsNodeOptions` before we actually have compiled it for which we need the `tsNodeOptions`. A typical chicken&egg problem 😄 It seems to be that the compiler options can be passed in as environment variables. I would suggest to keep things as they are and instead add some documentation as to how to modify these options.

We have received a lot of comments about how user can use imports in their config file and this solution should make this very easy. So I would love to be able to use TypeScript in the config without having to set up a special way to call wdio.

File: packages/wdio-types/src/index.ts
Comment (r4j4h) [reply packages/wdio-types/src/index.ts:24]: Great suggestion, that is good to know 👍

File: packages/wdio-config/src/utils.ts
Comment (r4j4h) [reply packages/wdio-config/src/utils.ts]: Great call, was also able to get rid of that "as any" 💃

File: packages/wdio-config/src/lib/ConfigParser.ts
Comment (r4j4h) [reply packages/wdio-config/src/lib/ConfigParser.ts:93]: Oooh, using environment variables is a great idea! Confirmed it works on our end too. 🎉 

OK I want to make sure I understand the new direction, are we saying:

1) Remove the new tsNodeOpts config stuff and add docs instructing to use environment variables for any specific config? https://github.com/TypeStrong/ts-node#cli-and-programmatic-options seems to give one for each so this should be sufficient.

Or are we saying:

2) Keep the new tsNodeOpts config stuff and also add docs letting people know about both options.

Comment (BorisOsipov) [reply packages/wdio-config/src/lib/ConfigParser.ts:93]: Environment variables can be pain in daily tests development routines. I don't like the idea when I need set env variable before I write\debug new test when I run `yarn run wdio --spec` every day.
It could sound crazy but I would like to have another config file e.g. `.wdiocompile` to control wdio autocompile behavior

File: CONTRIBUTING.md
Comment (r4j4h) [inline CONTRIBUTING.md:143]: I added a line about changing host or port. I can remove if you don't like it. I run as much of my development environment in containers and so the default localhost didn't work and I needed to dig into docusaurus to learn how to configure it and figured this might save other people that same time

File: packages/wdio-config/src/lib/ConfigParser.ts
Comment (r4j4h) [reply packages/wdio-config/src/lib/ConfigParser.ts:93]: Thanks for joining in! :)

.wdiocompile might be a good idea to also solve the chicken-and-egg config parsing problem too.

Another option would be to nested tsNodeOpts under autoCompileOpts, but I like how a dedicated config file would solve the parsing issue of the rest of the config, just wanted to mention it as an option though

Comment (christian-bromann) [reply packages/wdio-config/src/lib/ConfigParser.ts:93]: I think we should advocate both options. The important piece is that people can write TS in their config file and have the ability to add type safety to it. So maybe the tsconfig options are not really necessary for just compiling the config but when running the tests. To give more context: the config is read 2x:

1. in the launcher (main process) to read services and run the onPrepare and onComplete hooks
2. within the worker process to establish the framework, reporters and everything else

@r4j4h if you could find out whether you need the ts-node options also for just compiling the config, that would be great. Otherwise we have to be clear in the docs that only environment variables can be used if you have the need to set these configs.

Comment (r4j4h) [reply packages/wdio-config/src/lib/ConfigParser.ts:93]: @christian-bromann Great distinction. I would hazard a guess that it is only when running the tests, as it is after chromedriver is initialized:

```
2021-02-01T18:52:17.624Z DEBUG @wdio/config:utils: Found 'ts-node' package, auto-compiling TypeScript files

Execution of 1 spec files started at 2021-02-01T18:52:17.831Z

Starting ChromeDriver 87.0.4280.88 (89e2380a3e36c3464b5dd1302349b1382549290d-refs/branch-heads/4280@{#1761}) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
[1612205537.997][SEVERE]: bind() failed: Cannot assign requested address (99)
[0-0] RUNNING in chrome - /features/testing.feature
```

I am making pretty good progress toward .wdiocompile as well, for what that is worth

Comment (christian-bromann) [reply packages/wdio-config/src/lib/ConfigParser.ts:93]: Great! So we can keep the `tsNodeOpts` in our WebdriverIO Testrunner config and add to the docs that if people are interested adding specific compile options for the config they can through environment variables.

**Review by christian-bromann — COMMENTED** [review]
So I think two parts are missing here:

- move the transpiler call (`!loadTypeScriptCompiler(this._config.tsNodeOpts) && !loadBabelCompiler()`) above the line where we initially require the config
- add some documentation how to apply transpiler option through environment variables

Let me know if I can help with any of these.

**Comment by r4j4h** [comment]
Thanks, that seems quite simple. While adding the .wdiocompile file support, I've been adding to the test suite and increasing the testability of the ConfigParser, but I can probably split those out into separate MRs to get this one in quicker. Do you have a preference?

**Comment by r4j4h** [comment]
Also, by

> move the transpiler call (!loadTypeScriptCompiler(this._config.tsNodeOpts) && !loadBabelCompiler()) above the line where we initially require the config

Do you mean before we require the config file, or immediately after requiring it? I'm just a bit confused if this doesn't still break that situation that someone is using typescript in their config file? If this is fine to not customize ts-node here b/c it won't affect when it runs the tests, don't we still need to have it customizable somewhere before it runs the tests with it?

**Comment by christian-bromann** [comment]
> then parse them into as args into the worker

It might be necessary here to parse the ts-node config into the worker process as args and then be read by the `ConfigParser` before actually requiring the config file again.

**Comment by christian-bromann** [comment]
Yeah, the idea is to read the tsnode parameters when we read the config file in the testrunner process. Then pass these information along when the worker is being created so that we can apply them when we read the config in the work process. The config file is unlikely to have complex code in it and should be able to be compiled without a `tsconfig.json` so initially the testrunner reads these values and applies them in the worker.

**Comment by r4j4h** [comment]
OK, do let me know if there was already a way to distinguish between the launcher/runner context, but this works locally! Still need to update the docs

**Comment by r4j4h** [comment]
Have everything working with tests and even the .wdiocompile although we can remove that if you don't like it. Haven't pushed those parts because they refactor and would add noise.

Running into an error in the runner tests that I'm not sure of the direct cause of. At one point I had the enum in types, but seemed to be able to move it into config, but perhaps this is why I originally moved it to types and forgot?

```
    TypeError: Cannot read property 'Runner' of undefined

      62 | 
      63 | export default class Runner extends EventEmitter {
    > 64 |     private _configParser = new ConfigParser(ConfigEnvironmentContext.Runner)
```

What's odd is it is JUST the tests. I can build them fine and they work in the project fine. If you have any ideas it would be a big help!

**Comment by r4j4h** [comment]
Noticed the pipeline gave a much more helpful error:

```
Error: packages/wdio-config/src/lib/ConfigParser.ts(67,59): error TS2339: Property 'tsNodeOpts' does not exist on type 'Pick<Testrunner, "onPrepare" | "onWorkerStart" | "onComplete" | "onReload" | "before" | "beforeCommand" | "beforeHook" | "beforeSession" | "beforeSuite" | "beforeTest" | ... 56 more ... | "autoCompileOpts">'.
```

I think I missed one of the changes with the types so I went on and pushed everything up, sorry if it's a lot to look at

**Comment by r4j4h** [comment]
Hmm apparently it didn't give that error in the pipelines. The linter got mad at a test line I added though 😊

**Comment by r4j4h** [comment]
Awesome, well I'm exhausted for the night but it looks like it got way further, now the cli tests are yelling about the same thing. Will pick back up tomorrow

**Comment by r4j4h** [comment]
Nope, I don't think it got further, it's just doing the same thing as on my stack

File: packages/wdio-config/src/constants.ts
Comment (christian-bromann) [inline packages/wdio-config/src/constants.ts:98]: I don't think such a file is needed. If we can compile the config in the runner process and only need the compiler options in the worker we can just pass along the options from the runner process into the worker. Here is how this would go:

- the CLI launcher reads the config initially (https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-cli/src/launcher.ts#L58-L60), we can allow users to pass in some CLI args as you have implemented it already to allow modify compile options for that part. My assumption tho is that we don't need them normally there so compiling the config file should be fine
- we can now access the compile options defined in the config and pass them into the worker as [args](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-cli/src/launcher.ts#L377)
- these args are then available in the [runner](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-runner/src/index.ts#L86) and we can pass the compile options along when we [read the config file](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-runner/src/index.ts#L95)

That should make this a bit easer, what do you think?

Comment (christian-bromann) [reply packages/wdio-config/src/constants.ts:98]: This would also make the `ConfigEnvironmentContext` obsolete.

File: packages/wdio-types/src/Options.ts
Comment (christian-bromann) [inline packages/wdio-types/src/Options.ts:412]: I like this idea of putting Babel and TS options into one property. It can be also nicely extended for other purposes in the future 👍

**Review by christian-bromann — COMMENTED** [review]
Great progress 👍  Added some comments with optimisation ideas. Let me know if I can help with this. Feel free to book a [Open Office Hour](https://webdriver.io/blog/2020/07/01/office-hours.html) slot if you want to sync.

File: packages/wdio-config/package.json
Comment (r4j4h) [reply packages/wdio-config/package.json:35]: I used it in configparser.test.ts to simulate the glob behavior

File: packages/wdio-config/src/constants.ts
Comment (r4j4h) [reply packages/wdio-config/src/constants.ts]: Good call. Are you sure we need to pass along through the args since it re-loads the file anyway? I have it currently doing that and it seems to work fine, but I'm totally open to what you are asking just curious if we even need to at this point :)

**Comment by r4j4h** [comment]
I thought my changes to the tests would make things faster and easier to maintain but apparently I've caused a memory leak or something and the smoke tests are running out of memory :(

File: packages/wdio-types/src/Options.ts
Comment (r4j4h) [reply packages/wdio-types/src/Options.ts:412]: I'm glad you like this idea 🎉

File: website/docs/AutoCompile.md
Comment (christian-bromann) [inline website/docs/AutoCompile.md:3]: I wonder if it makes more sense to just mention these auto compile options in the Babel/TypeScript pages rather than having a dedicated page for it.

File: packages/wdio-config/src/lib/ConfigParser.ts
Comment (christian-bromann) [inline packages/wdio-config/src/lib/ConfigParser.ts:91]: What is the purpose of those? If it is for testing purposes we could just use the mock features of jest, no? It seems they create a lot of overhead, especially when calling:

```js
exclude = ConfigParser.getFilePaths(capExclude, undefined, this._pathService)
```

**Review by christian-bromann — COMMENTED** [review]
mhm .. interesting error. Are the smoke test working for you locally?

**Comment by r4j4h** [comment]
Yeah, locally it worked fine for me with the launcher settings being passed because that prevented ts-node from being loaded twice, but because I didn't quite follow the args approach and loaded ts-node twice in practice it works but the smoke tests blow up. The only thing messed up with the launcher settings is the TypeScript enum doesn't seem to carry over in the tests, even though it as well also works in practice.

I will look more into using the args as you described instead of my launcher settings as that should also work, even if it's just moving the launcher setting to the arg..

File: packages/wdio-config/src/lib/ConfigParser.ts
Comment (r4j4h) [reply packages/wdio-config/src/lib/ConfigParser.ts:91]: Yes they are for treating the file system as repositories. I desired testing different config files without needing to have more in the fixtures directory when all that changes is one property in them. In an earlier test I read the current fixture file into memory and modified it afterward, but it was feeling hacky.

As far as overhead I am not quite sure how that would be, it is passing a reference around to an object yes, but that object is running literally the same code as before. 

I think the mock features of jest would also work. I wanted to avoid coupling further with the implementation and mocking the process objects and felt capturing the abstract logic was cleaner. Both should absolutely work though, the big difference is where the hook for the tests happens: in application code as an interface instead of calling require, or in Node.js require infrastructure code providing mocked require responses for the code under test.

**Comment by r4j4h** [comment]
Glad we have the smoke tests. https://github.com/TypeStrong/ts-node/issues/409 indicates double loading ts-node is fraught with problems. That thread also makes me want to add a global killswitch in autoCompileOpts that just opts out of all of it for users who want that. What do you think?

File: packages/wdio-config/src/constants.ts
Comment (r4j4h) [reply packages/wdio-config/src/constants.ts]: Running into a problem with the config lifecycle where the current autocompile happens during `addConfigFile` calls and the args come in through `merge` calls which currently always happen sequentially after addConfigFile calls. These only seem to be separate because the runner shuts down if addConfigFile throws.

If we could combine them then we could check to see if the args has a autoCompileOpts and use that in the same way `ConfigEnvironmentContext` was being used to prevent double auto-compiler loads.

The reason `ConfigEnvironmentContext` worked before was it allowed addConfigFile to run ts-node ASAP in the launcher and to not run at all in the runner.

Merging them would allow us to have only one autoCompile check and call, where with `ConfigEnvironmentContext` before I had two of them, but it changes the API surface around that try/catch a little.

TL;DR; addConfigFile wants to run ts-node ASAP, so without altering the code there is no way to prevent a double-load in the runner via the args because addConfigFile will have already loaded ts-node with defaults by the time the merge code could load it using the args. :(

Comment (r4j4h) [reply packages/wdio-config/src/constants.ts]: Not merging them is also an option, but would require adding a flag if not to ConfigParser to the function calls, effectively bringing back `ConfigEnvironmentContext` just maybe in another name :/

File: website/docs/AutoCompile.md
Comment (r4j4h) [reply website/docs/AutoCompile.md:3]: Since it's dedicated behavior I felt a sub-page under Test Runner section was a good fit. Especially in light of https://github.com/TypeStrong/ts-node/issues/409, but seeing as WebdriverIO only supports Babel and TypeScript and there are pages for them already I see it makes sense to consolidate the information in both of those pages and remove this one which just effectively will point to either of them.

File: packages/wdio-config/src/constants.ts
Comment (r4j4h) [reply packages/wdio-config/src/constants.ts]: Ah, I am moving towards a solution of exposing a third function on ConfigParser to trigger auto-compilation, and triggering it before feeding files or merging in the launcher, and triggering it after those steps in the runner. Seems to do the trick!

**Comment by r4j4h** [comment]
Similarly to when I tried to add the enum for context, the other projects' tests do not see the autoCompile function I added. :(

```
this.configParser.autoCompile is not a function
```

**Comment by r4j4h** [comment]
Wasn't able to solve the type problem, but I was able to debug it further. In each case it was getting a Mocked ConfigParser that did not have those functions on it. But when I checked the node_modules folders manually it was there. Somehow jest or something got confused or there must be a type copy step I missed, but since this change is orthogonal to all of those adding duck-typing worked to get the tests happy :)

File: packages/wdio-cli/src/launcher.ts
Comment (christian-bromann) [inline packages/wdio-cli/src/launcher.ts:59]: Would that ever be undefined?
```suggestion
        if (this.configParser.autoCompile) {
```

File: packages/wdio-cli/tests/watcher.test.ts
Comment (christian-bromann) [inline packages/wdio-cli/tests/watcher.test.ts:16]: same
```suggestion
            if (this.configParser.autoCompile) {
```

File: packages/wdio-runner/src/index.ts
Comment (christian-bromann) [inline packages/wdio-runner/src/index.ts:105]: same
```suggestion
        if (this._configParser.autoCompile) {
```

File: packages/wdio-config/package.json
Comment (christian-bromann) [reply packages/wdio-config/package.json:35]: Are these tests pushed? I don't see them 🤔

**Review by christian-bromann — COMMENTED** [review]
Some last comments

File: examples/wdio.conf.js
Comment (r4j4h) [inline examples/wdio.conf.js:186]: Ooopps, I missed this

File: packages/wdio-config/src/lib/ConfigParser.ts
Comment (r4j4h) [inline packages/wdio-config/src/lib/ConfigParser.ts:125]: I actually meant to leave this out, as this is a change, but I noticed it was used elsewhere and seemed like it might be safer. Wanted to highlight it to make sure you saw :)

**Comment by r4j4h** [comment]
I think this is ready to go. It's a lot of commits though, would you like me to rebase it?

**Review by christian-bromann — APPROVED** [review]
This is excellent. I haven't seen such an high quality first time contribution for a long time. Thank you so much for all your effort and time 👍 

Do you have anything left to do here or can I go ahead and merge it?

File: packages/wdio-config/src/lib/ConfigParser.ts
Comment (christian-bromann) [reply packages/wdio-config/src/lib/ConfigParser.ts]: 👍

**Comment by r4j4h** [comment]
Thank you for working with me to get it such a nice state. It was my pleasure!

Nope I think we're ready to merge if you are!

**Comment by r4j4h** [comment]
Hah, I am so weird now that I type that I want to rebase it lol, give me a few more moments please it shouldn't take long

**Comment by r4j4h** [comment]
OK Now we should be ready to merge! :) sorry about that

**Comment by christian-bromann** [comment]
Thank you @r4j4h .. hope to see more of these excellent PRs in the future 😊

**Pull request merged**
